[Merged by Bors] - Add getting started section (#246)
stackabletech/superset-operator: docs/modules/getting_started/pages/first_steps.adoc

= First steps

> After you went through the xref:installation.adoc[] of the Operator and its dependencies, you will now go through the steps to set up a Superset instance and connect to the instance afterwards.

Reviewer: ```suggestion
Once you have followed the steps in  xref:installation.adoc[] for the Operator and its dependencies, you will now go through the steps to set up and connect to a Superset instance.
```

== Database for the Superset metadata

> The metadata of Superset (slices, connections, tables, dashboards, ...) is stored in an SQL database.

Reviewer: ```suggestion
Superset metadata (slices, connections, tables, dashboards etc.) is stored in an SQL database.
```

For testing purposes, you can spin up a PostgreSQL database with the following commands:

[source,bash]
include::example$code/getting_started.sh[tag=add-bitnami-repo]

[source,bash]
include::example$code/getting_started.sh[tag=install-bitnami-psql]

== Secret with Superset credentials

> A secret with the necessary credentials must be created, this entails database connection credentials as well as an admin account for Superset itself. Create a file called `superset-credentials.yaml`:

Reviewer: ```suggestion
A secret with the necessary credentials must be created: this contains database connection credentials as well as an admin account for Superset itself. Create a file called `superset-credentials.yaml`:
```

[source,yaml]
include::example$code/superset-credentials.yaml[]

And apply it:

[source,bash]
include::example$code/getting_started.sh[tag=apply-superset-credentials]

The `connections.secretKey` will be used for securely signing the session cookies and can be used
for any other security related needs by extensions. It should be a long random string of bytes.

`connections.sqlalchemyDatabaseUri` must contain the connection string to the SQL database storing
the Superset metadata.

The `adminUser` fields are used to create an admin user.
Please note that the admin user will be disabled if you use a non-default authentication mechanism like LDAP.

== Creation of a Superset node

A Superset node must be created as a custom resource, create file called `superset.yaml`:

[source,yaml]
include::example$code/superset.yaml[]

And apply it:

[source,bash]
include::example$code/getting_started.sh[tag=apply-superset-cluster]

`metadata.name` contains the name of the Superset cluster.

The label of the Docker image provided by Stackable must be set in `spec.version`.

Please note that the version you need to specify is not only the version of Apache Superset which

> you want to roll out, but has to be amended with a Stackable version as shown. This Stackable

Reviewer: ```suggestion
you want to roll out, but has to also include the Stackable version as shown. This Stackable
```

version is the version of the underlying container image which is used to execute the processes. For
a list of available versions please check our
https://repo.stackable.tech/#browse/browse:docker:v2%2Fstackable%2Fsuperset%2Ftags[image registry].
It should generally be safe to simply use the latest image version that is available.

`spec.statsdExporterVersion` must contain the tag of a statsd-exporter Docker image in the Stackable repository.

The previously created secret must be referenced in `spec.credentialsSecret`.

The `spec.loadExamplesOnInit` key is optional and defaults to `false`, it can be set to `true` to load example data into Superset when the database is initialized.

The `rowLimit` configuration option defines the row limit when requesting chart data.

=== Initialization of the Superset database

> The first time the cluster is created, the operator creates a `SupersetDB` resource with the same name as the cluster.  It ensures that the database is initialized (schema created, admin user created).

Reviewer: ```suggestion
The first time the cluster is created, the operator creates a `SupersetDB` resource with the same name as the cluster.  It ensures that the database is initialized (with the creation of a schema and admin user).
```

A Kubernetes job is created which starts a pod to initialize the database. This can take a while.

You can use kubectl to wait on the resource:

[source,bash]
include::example$code/getting_started.sh[tag=wait-supersetdb]

== Using Superset

When the Superset node is created and the database is initialized, Superset can be opened in the
browser.

The Superset port which defaults to `8088` can be forwarded to the local host:

[source,bash]
include::example$code/getting_started.sh[tag=port-forwarding]

> Then it can be opened in the browser with `http://localhost:8088`.

Reviewer: This doesn't work for me. The output of the script is fine:
```
Starting port-forwarding of port 8088
Checking if web interface is reachable ...
Web interface reachable!
```
but opening the url in the browser yields nothing: the port forward seems to not be persisted when the script finishes. If I create it manually (again) in k9s then everything is fine.

Author: Yes, this is on purpose, the script cleans up after itself. Look in line 78 at the `trap` command. A `kill` command is executed on the EXIT signal of the script. If the user just pastes the single line from the document, it will stay open.

Reviewer: ah, ok - that was a misunderstanding on my part. So the intention is **not** that the user can/should execute the getting_started.sh script, but that we use it as a) a sanity check that everything works, and b) a source for extracting code snippets to eliminate redundancy? Maybe we should make it clear that the script can be used, but does its own cleanup.

Author: Yes, it's not linked anywhere and we don't want the user to use it. This is mostly to make sure the stuff in the docs works (so as a test) and possibly to use for us internally.

I'll add some comments on that it does its own cleanup and what it is intended for.

Reviewer: Thanks!
